### ISSUE TITLE
memory leak on sqlite

### ISSUE THREAD

**Issue by shuffleyxf** [sqlobject/sqlobject#186]
It seems that when accessing SQLite in file mode, SQLObject caches the connection for each thread and does not release it even if the thread has ended. The following code reproduces the issue.
```
from flask import Flask
from sqlobject import *

class Person(SQLObject):
     fname = StringCol()
     mi = StringCol(length=1, default=None)
     lname = StringCol()


app = Flask('app')


@app.route('/person')
def query_person():
    return {"result": do_select()}

def do_select():
    list(Person.selectBy())

if __name__ == '__main__':
    sqlhub.processConnection = connectionForURI('sqlite:/:test.db')
    Person.createTable(ifNotExists=True)
    app.run()
```
Every request generates a new connection.

**Comment by phdru** [comment]
Are you sure it's not a problem with Flask? IWBN if you create an SQLObject-only test case.

**Comment by shuffleyxf** [comment]
The following code also reproduces the issue.
```
import time
from threading import Thread

from sqlobject import *

class Person(SQLObject):
     fname = StringCol()
     mi = StringCol(length=1, default=None)
     lname = StringCol()

def multi_thread_test():
    while True:
        Thread(target=do_select).start()
        time.sleep(0.1)

def do_select():
    list(Person.selectBy())


if __name__ == '__main__':
    sqlhub.processConnection = connectionForURI('sqlite:/:test.db')
    Person.createTable(ifNotExists=True)
    multi_thread_test()
```
When you run this code, you will notice a rapid increase in memory. I believe the cause is increasing SQLite connections.

**Comment by phdru** [comment]
Tnx, that's better! Will look into it.

**Comment by phdru** [comment]
This is the command I used to see the process status:
```sh
$ ps auxwww | grep test | grep -v grep
phd      14912  0.9  0.3 106836 24828 pts/6    SN+  20:19   0:03 python3 ./test-issue186.py
```
I ran the command for 10 minutes but neither virtual size (106836) nor RSS (24828) increased. No leaks for me.

I use Debian 12 Linux, Python 3.11, SQLObject 3.10.2 (the latest).

**Comment by shuffleyxf** [comment]
I reproduced the issue with Windows 10, Python 3.10.5, SQLObject 3.10.2. I also tested it with Ubuntu 22, Python 3.10.5, SQLObject 3.10.2, and there was no memory leak. Through further testing, I found that on Ubuntu 22, Python reuses thread ID when creating new threads, whereas on Windows 10, it does not. 
The result of the following code demonstrates that the memory is not released after the threads end.
```
import time
import threading
import psutil
from threading import Thread

from sqlobject import *


class Person(SQLObject):
     fname = StringCol()
     mi = StringCol(length=1, default=None)
     lname = StringCol()


def display_memory():
    print(f'rss: {psutil.Process().memory_info().rss / 1024}KB')


def multi_thread_test(target):
    display_memory()

    thread_num = 50
    threads = []
    for x in range(thread_num):
        t = Thread(target=target)
        threads.append(t)
        t.start()

    for t in threads:
        t.join()

    display_memory()


def do_select():
    time.sleep(1)
    list(Person.selectBy())


def do_nothing():
    time.sleep(1)


if __name__ == '__main__':
    sqlhub.processConnection = connectionForURI('sqlite:./test.db')
    Person.createTable(ifNotExists=True)
    multi_thread_test(do_nothing)
    multi_thread_test(do_select)
    time.sleep(600)

```
In fact, I don't think this would be a problem on Linux, but it can easily have an impact on Windows.

**Comment by phdru** [comment]
> … I found that on Ubuntu 22, Python reuses thread ID when creating new threads, whereas on Windows 10, it does not.

Interesting observation, thank you! Food for thought, for sure.

It'd be harder to debug as I almost never used Windows and I'm not sure my current installation of QEMU works. While I'm recovering it can you teach me how to monitor memory usage of a single process?

**Comment by shuffleyxf** [comment]
Sure. On Windows, you can use the built-in Task Manager to monitor the memory usage of a process.

**Comment by phdru** [comment]
> Sure. On Windows, you can use the built-in Task Manager to monitor the memory usage of a process.

Tnx!

**Comment by phdru** [comment]
My preliminary thought is: SQLObject alone cannot fix that. Multithreading applications should cooperate. Python doesn't provide a way to callback on thread end. The application should close the connection at the end of every thread.
```python
def do_select():
    list(Person.selectBy())
    sqlhub.processConnection.close()  # or Person._connection.close(), it's the same
```
I can create a wrapper that calls `.close()` at the thread end but applications need to use it explicitly. It could be something like
```python
from sqlobject.sqliteconnection import close_connection

def multi_thread_test():
    while True:
        Thread(target=close_connection(do_select)).start()
```

**Comment by shuffleyxf** [comment]
Great. I am wondering if there is any solution that can solve this problem without involving any user code. What do you think about releasing useless connections when a new connection is created? I think the following code can get the IDs of alive threads.
```
alive_thread_id_list = [t.ident for t in threading.enumerate()]
```
Is it possible to detect whether each cached connection corresponds to a live thread, and if not, release that connection? Perhaps we could also add a threshold to reduce the frequency of this operation.

**Comment by phdru** [comment]
> Great. I am wondering if there is any solution that can solve this problem without involving any user code. What do you think about releasing useless connections when a new connection is created? I think the following code can get the IDs of alive threads.
> 
> ```
> alive_thread_id_list = [t.ident for t in threading.enumerate()]
> ```
> 
> Is it possible to detect whether each cached connection corresponds to a live thread, and if not, release that connection? Perhaps we could also add a threshold to reduce the frequency of this operation.

The approach is worth trying. Gonna explore…

**Comment by phdru** [comment]
I revived my QEMU, installed Windows 7 and Python 3.7, created a virtual env and ran your second program. Ouch, there was leak about 100 Kb/sec.

I added `Person._connection.close()` at the end of `do_select` and the problem became much less painful but didn't disappear. There is still leak about one Kb per second. 100 times less but still leak.

Gonna try enumerating threads…

**Comment by phdru** [comment]
I believe the problem is fixed. Please test and report.

**Comment by shuffleyxf** [comment]
I ran the second program with Windows 10, and it still leaks. It appears that `SQLiteConnection._thread_pool` still maintains the connection.

**Comment by phdru** [comment]
Let's clear `_threadPool` — is it better?

**Comment by phdru** [comment]
I removed SQLObject from the program so the code becomes this:
```python
#! /usr/bin/env python3

import time
from threading import Thread

def multi_thread_test():
    while True:
        Thread(target=do_select).start()
        time.sleep(0.1)

def do_select():
    time.sleep(0.1)

if __name__ == '__main__':
    multi_thread_test()
```
I ran it for a few minutes — it leaks. Can you please confirm there is non-SQLObject-related leak?

**Comment by shuffleyxf** [comment]
I ran your program, no leaks for me, maybe you should try extending the duration of the first sleep.

**Comment by shuffleyxf** [comment]
> Let's clear `_threadPool` — is it better?

I tried modifying `SQLiteConnection.releaseConnection`, and my second program no longer leaks. The modified version is as follows.
```
    def releaseConnection(self, conn, explicit=False):
        if self._memory:
            return
        threadid = self._threadOrigination.get(id(conn))
        DBAPI.releaseConnection(self, conn, explicit=explicit)
        if (self._pool is not None and threadid
                and threadid not in self._threadPool):
            self._threadPool[threadid] = conn
        else:
            if self._pool and conn in self._pool:
                self._pool.remove(conn)
            if threadid:
                del self._threadOrigination[id(conn)]
                del self._threadPool[threadid]
            conn.close()
```

**Comment by phdru** [comment]
Ok, let's try this one. There is still some leak for me.

**Comment by phdru** [comment]
@shuffleyxf Any opinion on the commit 5aa8d77 ?

**Comment by phdru** [comment]
Any opinion on the commit fbff63a ?

**Comment by shuffleyxf** [comment]
Great. I think the bug is completely fixed.

**Comment by phdru** [comment]
Thanks for reporting!

**Comment by phdru** [comment]
Released 3.10.3 with your fixes. Thank you very much!
